ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,721 @@
+		scratch,
+		tries,
+		subTrieRootIndices,
+		subTriesNodeCount+1, // the counter is 1 more than the node count, because the first item is nil
+		writer)
+
+	if err != nil {
+		return 0, fmt.Errorf("could not store top level nodes: %w", err)
+	}
+
+	logger.Info().Msgf("top level nodes have been stored. top level node count: %v", topLevelNodesCount)
+
+	// write tries
+	err = storeTries(scratch, tries, topLevelNodeIndices, writer)
+	if err != nil {
+		return 0, fmt.Errorf("could not store trie root nodes: %w", err)
+	}
+
+	// write checksum
+	checksum, err := storeTopLevelTrieFooter(topLevelNodesCount, uint16(len(tries)), writer)
+	if err != nil {
+		return 0, fmt.Errorf("could not store footer: %w", err)
+	}
+
+	return checksum, nil
+}
+
+func createSubTrieRoots(tries []*trie.MTrie) [subtrieCount][]*node.Node {
+	var subtrieRoots [subtrieCount][]*node.Node
+	for i := 0; i < len(subtrieRoots); i++ {
+		subtrieRoots[i] = make([]*node.Node, len(tries))
+	}
+
+	for trieIndex, t := range tries {
+		// subtries is an array with subtrieCount trie nodes
+		// in breadth-first order at subtrieLevel of the trie `t`
+		subtries := getNodesAtLevel(t.RootNode(), subtrieLevel)
+		for subtrieIndex, subtrieRoot := range subtries {
+			subtrieRoots[subtrieIndex][trieIndex] = subtrieRoot
+		}
+	}
+	return subtrieRoots
+}
+
+// estimateSubtrieNodeCount estimate the average number of registers in each subtrie.
+func estimateSubtrieNodeCount(trie *trie.MTrie) int {
+	estimatedTrieNodeCount := 2*int(trie.AllocatedRegCount()) - 1
+	return estimatedTrieNodeCount / subtrieCount
+}
+
+// subTrieRootAndTopLevelTrieCount return the total number of subtrie root nodes
+// and top level trie nodes for given number of tries
+// it is used for preallocating memory for the map that holds all unique nodes in
+// all sub trie roots and top level trie nodoes.
+// the top level trie nodes has nearly same number of nodes as subtrie node count at subtrieLevel
+// that's it needs to * 2.
+func subTrieRootAndTopLevelTrieCount(tries []*trie.MTrie) int {
+	return len(tries) * subtrieCount * 2
+}
+
+func storeSubTrieConcurrently(
+	subtrieRoots [subtrieCount][]*node.Node,
+	estimatedSubtrieNodeCount int, // useful for preallocating memory for building unique node map when processing sub tries
+	subAndTopNodeCount int, // useful for preallocating memory for the node indices map to be returned
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+	nWorker uint,
+) (
+	map[*node.Node]uint64, // node indices
+	uint64, // node count
+	[]uint32, //checksums
+	error, // any exception
+) {
+	logger.Info().Msgf("storing %v subtrie groups with average node count %v for each subtrie", subtrieCount, estimatedSubtrieNodeCount)
+
+	if nWorker == 0 || nWorker > subtrieCount {
+		return nil, 0, nil, fmt.Errorf("invalid nWorker %v, the valid range is [1,%v]", nWorker, subtrieCount)
+	}
+
+	resultChs := make(chan chan *resultStoringSubTrie, nWorker-1)
```

Comment:
Although `nWorker-1` makes sense for single goroutine checkpointing, it isn't a good fit for multi-goroutine checkpointing.

For example, `nWorker` is 16  in `StoreCheckpointV6Concurrent()`, and it doesn't make sense to change it to 17.

To simplify, I think we can think of `nWorker` as extra workers.  It is 0 (default value) for single goroutine, and `16` for parallel.  So this block of code doesn't need to modify `nWorker`.

https://github.com/onflow/flow-go/blob/44e630a0064d68c793598ef1fd48a056681ce524/ledger/complete/wal/checkpoint_v6_writer.go#L46-L48